### PR TITLE
C++: Fix coroutine IR inconsistencies

### PR DESCRIPTION
**Pull request by jketema**

### CODE REVIEW

File: cpp/downgrades/7ff6a6e53dbcff09d1b9b758b594bc6d17366863/coroutine.ql [hunk 1]
```diff
@@ -0,0 +1,18 @@
+ class Function extends @function {
```
Comment (geoffw0) [inline cpp/downgrades/7ff6a6e53dbcff09d1b9b758b594bc6d17366863/coroutine.ql:1]:
```suggestion
class Function extends @function {
```

File: cpp/downgrades/7ff6a6e53dbcff09d1b9b758b594bc6d17366863/coroutine.ql [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+class Variable extends @variable {
+  string toString() { none() }
+}
+
+from Function func, Type traits, Variable handle, Variable promise
+where
+  coroutine(func, traits) and
+  coroutine_placeholder_variable(handle, 1, func) and
+  coroutine_placeholder_variable(promise, 2, func)
+select func, traits, handle, promise
```
Comment (geoffw0) [inline cpp/downgrades/7ff6a6e53dbcff09d1b9b758b594bc6d17366863/coroutine.ql:18]: I assume the point of this change is that new-style coroutines don't have to have exactly 1 handle and 1 promise.  Can they have zero now?  Can they have more than one?

If they have zero, I can see that this downgrade script is simply going to delete the coroutine, which actually seems fine - we wouldn't have been extracting anything sensible before.

If more than one, we're going to get something of a cartesian product or `coroutine` rows here.  I'm guessing we only expect small numbers, but I want to draw attention to this possible behaviour.

File: cpp/ql/test/library-tests/ir/ir/aliased_ssa_consistency.expected [hunk 3]
```diff
@@ -29,26 +29,4 @@ fieldAddressOnNonPointer
...
-| file://:0:0:0:0 | VariableAddress: (unnamed local variable) | Variable address instruction 'VariableAddress: (unnamed local variable)' has no associated variable, in function '$@'. | coroutines.cpp:99:21:99:40 | co_returnable_value co_yield_value_value(int) | co_returnable_value co_yield_value_value(int) |
-| file://:0:0:0:0 | VariableAddress: (unnamed local variable) | Variable address instruction 'VariableAddress: (unnamed local variable)' has no associated variable, in function '$@'. | coroutines.cpp:103:20:103:43 | co_returnable_void co_yield_and_return_void(int) | co_returnable_void co_yield_and_return_void(int) |
-| file://:0:0:0:0 | VariableAddress: (unnamed local variable) | Variable address instruction 'VariableAddress: (unnamed local variable)' has no associated variable, in function '$@'. | coroutines.cpp:108:21:108:45 | co_returnable_value co_yield_and_return_value(int) | co_returnable_value co_yield_and_return_value(int) |
```
Comment (geoffw0) [inline cpp/ql/test/library-tests/ir/ir/aliased_ssa_consistency.expected:53]: :tada:

**Review by geoffw0 — COMMENTED** [review]
One question...

File: cpp/downgrades/7ff6a6e53dbcff09d1b9b758b594bc6d17366863/coroutine.ql (hunk 2 above)
Comment (jketema) [reply cpp/downgrades/7ff6a6e53dbcff09d1b9b758b594bc6d17366863/coroutine.ql:18]: >I assume the point of this change is that new-style coroutines don't have to have exactly 1 handle and 1 promise. Can they have zero now? Can they have more than one?

They will have exactly one of each. The problem with the previous scheme is that it's not easy to add additional variables, which is what we need here. Adding an extra column to `coroutine` makes it difficult to write an upgrade script, as I don't have anything I can sensibly put in the additional column.

Comment (geoffw0) [reply cpp/downgrades/7ff6a6e53dbcff09d1b9b758b594bc6d17366863/coroutine.ql:18]: Makes sense, and nullifies my concern above - thanks.

**Review by geoffw0 — APPROVED** [review]

**Pull request merged**